### ISSUE TITLE
build R-INLA from source without binaries

### ISSUE THREAD

**Issue by deniskristak** [hrue/r-inla#22]
Hi,
In the directory build-user/linux/Makefile I see that some dependencies are being downloaded as binaries.
I'd like to install R-INLA from source, separately, without any binaries stepping in, while taking care of extensions myself (building them from source as well).
Is there a way to skip downloading all these dependency binary files? Also, is there a file which lists all dependencies and their versions (just like there are typically in conda-recipe/meta.yaml for anaconda or requirements.txt for pip)?

fyi, the purpose of this is to include possibility to build R-INLA with EasyBuild - see https://github.com/easybuilders/easybuild,
where we try to build everything from source - including dependencies.

thank you very much!

**Comment by hrue** [comment]
well... you can build it almost from source.   see hrue.r-inla-download.org/build64.tgz  which is for Fedora(33).

you have to use the taucs/amd/muparser versions of the libs in there

I set
    export INLA_TAG=Version_21.01.18
and do
    make update
then it builds it in local/bin

the metis5 library is not needed as its included in the pardiso lib, unless you're on Windows. 

you have to include src/libpardiso.o 

> $ diff inlaprog/Makefile inlaprog/GNUmakefile 
> 79c79
> < OBJ = src/inla.o src/dictionary.o  src/iniparser.o  src/strlib.o src/my-fix.o src/spde.o src/spde2.o src/spde3.o src/eval.o src/interpol.o src/re.o src/ar.o src/bessel.o src/pc-priors.o src/R-interface.o src/fgn.o src/quantile-regression.o src/my.o src/stochvol.o src/xerbla.o src/libpardiso.o src/tweedie.o
> ---
> > OBJ = src/inla.o src/dictionary.o  src/iniparser.o  src/strlib.o src/my-fix.o src/spde.o src/spde2.o src/spde3.o src/eval.o src/interpol.o src/re.o src/ar.o src/bessel.o src/pc-priors.o src/R-interface.o src/fgn.o src/quantile-regression.o src/my.o src/stochvol.o src/xerbla.o src/tweedie.o ##src/libpardiso.o
> 

if you do not link with the PARDISO library, and define -DNO_PARDISO_LIB when compiling.

you also need to set cpp var if you link with openblas or Intel MKL; see the Makefile. By default, it will compile both.

I hope this is useful. Its a little messy, and we should rewrite this build system. I know...

H

**Comment by boegel** [comment]
@hrue I'm working together with @deniskristak on this, and we would really like to get a full from-source build working, where we don't use any pre-built libraries. Most of the required dependencies (BLAS/LAPACK via OpenBLAS or MKL, SuiteSparse, METIS, GSL, libRmath, muParser) are already supported in [EasyBuild](https://easybuild.io), so we would use those installations (not the ones coming from Fedora or Ubuntu).

We have a lot of experience with complex installations that involve lots of dependencies, so I think we should be able to get this fully working (and make it easy for others to install `R-INLA` from source).

I think we're slowly figuring things out, but we need some help...
A couple of things we ran into aren't fully clear to use yet:

1) We plan to use a clone of this repository, at a specific tag (like [21.03.31](https://github.com/hrue/r-inla/releases/tag/Version_21.03.31)), as a starting point for the from-source installation, does that make sense?

2) It looks like the [build-user/linux/Makefile](https://github.com/hrue/r-inla/blob/devel/build-user/linux/Makefile) specifies to full install procedure, and we should mainly focus on the `make bin` part of it, which builds `taucs`, `fmesher`, `GMRFLib`, and finally `inla`. Most of those are basic `make`/`make install` procedures, so that shouldn't be too difficult... Is that correct?

3) For `taucs`, it seems like you're using a patched version, so we should also use the source tarball available in [extlibs](https://github.com/hrue/r-inla/tree/devel/extlibs)?

4) Do we also need to use the `muparser_v2_2_3-with-gamma.tgz` tarball in that same directory, or is it OK to use the latest `muparser` 2.2.5 ?

5) Can you clarify why `$INLA_TAG` needs to be set, and where that is picked up?

6) In which `Makefile` is it specified that R-INLA is built on top of OpenBLAS or MKL? We want to just build for one of those (at a time).

7) What's the relevance of the PARDISO library, how is R-INLA different when built with(out) it? And how does that affect the METIS dependency exactly?

8) How do you recommend we test the resulting R-INLA installation?

**Comment by hrue** [comment]
Hi.

my comments are below. If there is anything else, let me know

Best
Håvard


On Wed, 2021-04-07 at 03:20 -0700, Kenneth Hoste wrote:
@hrue I'm working together with @deniskristak on this, and we would really like to get a full from-source build working, where we don't use any pre-built libraries. Most of the required dependencies (BLAS/LAPACK via OpenBLAS or MKL, SuiteSparse, METIS, GSL, libRmath, muParser) are already supported in EasyBuild, so we would use those installations (not the ones coming from Fedora or Ubuntu).
We have a lot of experience with complex installations that involve lots of dependencies, so I think we should be able to get this fully working (and make it easy for others to install R-INLA from source).
I think we're slowly figuring things out, but we need some help...
A couple of things we ran into aren't fully clear to use yet:

   1. We plan to use a clone of this repository, at a specific tag (like 21.03.31), as a starting point for the from-source installation, does that make sense?

All tags like Version_21.03.31 defines a build of the package (or a try to in some cases) and this is the one you'll find the the R package INLA_21.03.31

   2. It looks like the build-user/linux/Makefile specifies to full install procedure, and we should mainly focus on the make bin part of it, which builds taucs, fmesher, GMRFLib, and finally inla. Most of those are basic make/make install procedures, so that shouldn't be too difficult... Is that correct?

I'm not sure it really updated, but I uploaded my build directory on my Fedora33 laptop to

hrue.r-inla-download.org/build64.tgz

and you can go from there.

at top, you'll make 'amd' 'muparser_v2_2_3' 'taucs-2.2--my-fix'

which is my patched libraries.

you can use a newer muparser I guess, I just added some more gamma functions to it (and I do not think its really in use...)

src/muParser.cpp- //---------------------------------------------------------------------------
src/muParser.cpp- // misc
src/muParser.cpp: value_type Parser::Lgamma(value_type v) { return lgamma((double)v); }
src/muParser.cpp: value_type Parser::Gamma(value_type v) { return tgamma((double)v); }
src/muParser.cpp: value_type Parser::Digamma(value_type v) { return digamma((double)v); }
src/muParser.cpp: value_type Parser::Trigamma(value_type v) { return trigamma((double)v); }
src/muParser.cpp-
src/muParser.cpp- value_type Parser::Exp(value_type v) { return MathImpl<value_type>::Exp(v); }
--
src/muParser.cpp- DefineFun(_T("max"), Max);
src/muParser.cpp-
src/muParser.cpp: DefineFun(_T("gamma"), Gamma);
src/muParser.cpp: DefineFun(_T("lgamma"), Lgamma);
src/muParser.cpp: DefineFun(_T("digamma"), Digamma);
src/muParser.cpp: DefineFun(_T("trigamma"), Trigamma);


   3. For taucs, it seems like you're using a patched version, so we should also use the source tarball available in extlibs?

yes


   4. Do we also need to use the muparser_v2_2_3-with-gamma.tgz tarball in that same directory, or is it OK to use the latest muparser 2.2.5 ?

see above

   5. Can you clarify why $INLA_TAG needs to be set, and where that is picked up?

I do from shell

export INLA_TAG="Version_21.03.31"
make update

which will 'git pull' and switch to that tag, and set some GIT cpp variables which is added using 'gcc -D...'

you can do INLA_TAG=devel  f.example




   6. In which Makefile is it specified that R-INLA is built on top of OpenBLAS or MKL? We want to just build for one of those (at a time).

If you will link with OpenBLAS then compile with

-DINLA_LINK_WITH_OPENBLAS

and

-DINLA_LINK_WITH_MKL

for MKL.  There is a function call in openmp.c hat depends on this, and for simplicity I just compile twice. Maybe there is a way to get around it, but...



   7. What's the relevance of the PARDISO library, how is R-INLA different when built with(out) it? And how does that affect the METIS dependency exactly?

tricky. PARDISO is a by far better parallel sparse solver, which require a license to run, which is free for academic++, so I only have access to
a prebuildt version.

The TAUCS library was stopped develloped in about 2003 or so, and is the default one. The PARDISO is faaar better and the only option that scales well to larger models, and runs in parallel etc

by default, the library will not link with PARDISO and a 'fake' PARDISO library is included to account for it, inlaprog/src/libpardiso.c

so I override 'Makefile' with a local copy 'GNUmakefile' to comment out this file from the build:

$ pwd
~/builds/build64/r-inla/inlaprog

***@***.******@***.***>([tag-Version_21.03.31]): ~/builds/build64/r-inla/inlaprog

$ diff GNUmakefile Makefile
79c79
< OBJ = src/inla.o src/dictionary.o src/iniparser.o src/strlib.o src/my-fix.o src/spde.o src/spde2.o src/spde3.o src/eval.o src/interpol.o src/re.o src/ar.o src/bessel.o src/pc-priors.o src/R-interface.o src/fgn.o src/quantile-regression.o src/my.o src/stochvol.o src/xerbla.o src/tweedie.o ##src/libpardiso.o
---
> OBJ = src/inla.o src/dictionary.o src/iniparser.o src/strlib.o src/my-fix.o src/spde.o src/spde2.o src/spde3.o src/eval.o src/interpol.o src/re.o src/ar.o src/bessel.o src/pc-priors.o src/R-interface.o src/fgn.o src/quantile-regression.o src/my.o src/stochvol.o src/xerbla.o src/libpardiso.o src/tweedie.o



PARDISO comes with its own version of the METIS library, so linking with that one does not require linking with METIS, which you otherwise must link with.


   8. How do you recommend we test the resulting R-INLA installation?

do some of the demos in the R-INLA package, like

inla.setOption(inla.call="path to your binary")
demo(Bym)

etc




--

Håvard Rue
Professor of Statistics
Statistics Program, CEMSE Division
King Abdullah University of Science and Technology
Thuwal 23955-6900, Saudi Arabia

Email: ***@***.***
Office: +966 (0)12 808 0640
Mobile: +966 (0)54 470 0421
Research group: bayescomp.kaust.edu.sa
R-INLA project: www.r-inla.org
Zoom: kaust.zoom.us/my/haavard.rue

**Comment by hrue** [comment]
I also link with jemalloc library, which has a HUGE impact for some
models on the speed. 




On Wed, 2021-04-07 at 20:05 +0300, Haavard Rue wrote:
> 
> Hi.
> 
> my comments are below. If there is anything else, let me know
> 
> Best
> Håvard
> 
> 
> On Wed, 2021-04-07 at 03:20 -0700, Kenneth Hoste wrote:
> 
> 
> 
> 
> All tags like Version_21.03.31 defines a build of the package (or a
> try to in some cases) and this is the one you'll find the the R
> package INLA_21.03.31
> 
> 
> 
> I'm not sure it really updated, but I uploaded my build directory on
> my Fedora33 laptop to
> 
> 	hrue.r-inla-download.org/build64.tgz
> 
> and you can go from there.
> 
> at top, you'll make 'amd' 'muparser_v2_2_3' 'taucs-2.2--my-fix'
> 
> which is my patched libraries.
> 
> you can use a newer muparser I guess, I just added some more gamma
> functions to it (and I do not think its really in use...)
> 
> src/muParser.cpp- //--------------------------------------------------
> -------------------------
> src/muParser.cpp- // misc
> src/muParser.cpp: value_type Parser::Lgamma(value_type v) { return
> lgamma((double)v); }
> src/muParser.cpp: value_type Parser::Gamma(value_type v) { return
> tgamma((double)v); }
> src/muParser.cpp: value_type Parser::Digamma(value_type v) { return
> digamma((double)v); }
> src/muParser.cpp: value_type Parser::Trigamma(value_type v) { return
> trigamma((double)v); }
> src/muParser.cpp-
> src/muParser.cpp- value_type Parser::Exp(value_type v) { return
> MathImpl<value_type>::Exp(v); }
> --
> src/muParser.cpp- DefineFun(_T("max"), Max);
> src/muParser.cpp-
> src/muParser.cpp: DefineFun(_T("gamma"), Gamma);
> src/muParser.cpp: DefineFun(_T("lgamma"), Lgamma);
> src/muParser.cpp: DefineFun(_T("digamma"), Digamma);
> src/muParser.cpp: DefineFun(_T("trigamma"), Trigamma);
> 
> 
> 
> 
> yes
> 
> 
> 
> 
> see above
> 
> 
> 
> I do from shell
> 
> 	export INLA_TAG="Version_21.03.31"
> 	make update
> 
> which will 'git pull' and switch to that tag, and set some GIT cpp
> variables which is added using 'gcc -D...' 
> 
> you can do INLA_TAG=devel  f.example
> 
> 
> 
> 
> 
> 
> If you will link with OpenBLAS then compile with
> 
> -DINLA_LINK_WITH_OPENBLAS
> 
> and
> 
> -DINLA_LINK_WITH_MKL
> 
> for MKL.  There is a function call in openmp.c hat depends on this,
> and for simplicity I just compile twice. Maybe there is a way to get
> around it, but...
> 
> 
> 
> 
> 
> tricky. PARDISO is a by far better parallel sparse solver, which
> require a license to run, which is free for academic++, so I only have
> access to 
> a prebuildt version.
> 
> The TAUCS library was stopped develloped in about 2003 or so, and is
> the default one. The PARDISO is faaar better and the only option that
> scales well to larger models, and runs in parallel etc
> 
> by default, the library will not link with PARDISO and a 'fake'
> PARDISO library is included to account for it,
> inlaprog/src/libpardiso.c
> 
> so I override 'Makefile' with a local copy 'GNUmakefile' to comment
> out this file from the build:
> 
> $ pwd
> ~/builds/build64/r-inla/inlaprog
> 
> ***@***.***([tag-Version_21.03.31]): ~/builds/build64/r-
> inla/inlaprog
> 
> $ diff GNUmakefile Makefile
> 79c79
> < OBJ = src/inla.o src/dictionary.o src/iniparser.o src/strlib.o
> src/my-fix.o src/spde.o src/spde2.o src/spde3.o src/eval.o
> src/interpol.o src/re.o src/ar.o src/bessel.o src/pc-priors.o src/R-
> interface.o src/fgn.o src/quantile-regression.o src/my.o
> src/stochvol.o src/xerbla.o src/tweedie.o ##src/libpardiso.o
> ---
> > OBJ = src/inla.o src/dictionary.o src/iniparser.o src/strlib.o
> > src/my-fix.o src/spde.o src/spde2.o src/spde3.o src/eval.o
> > src/interpol.o src/re.o src/ar.o src/bessel.o src/pc-priors.o src/R-
> > interface.o src/fgn.o src/quantile-regression.o src/my.o
> > src/stochvol.o src/xerbla.o src/libpardiso.o src/tweedie.o
> 
> 
> 
> PARDISO comes with its own version of the METIS library, so linking
> with that one does not require linking with METIS, which you otherwise
> must link with.
> 
> 
> 
> 
> do some of the demos in the R-INLA package, like
> 
> 	inla.setOption(inla.call="path to your binary")
> 	demo(Bym)
> 
> etc
> 
> 
> 

-- 
Håvard Rue
Professor of Statistics
Statistics Program, CEMSE Division
King Abdullah University of Science and Technology
Thuwal 23955-6900, Saudi Arabia

***@***.***
Office: +966 (0)12 808 0640  
Mobile: +966 (0)54 470 0421
Research group: bayescomp.kaust.edu.sa  
R-INLA project: www.r-inla.org
Zoom: kaust.zoom.us/my/haavard.rue

**Comment by boegel** [comment]
If `build-user/linux/Makefile` is not the latest version of the build procedure, is there another script/Makefile included in the repository that describes all the steps?

We can check the contents of hrue.r-inla-download.org/build64.tgz  for inspiration (if it includes useful stuff next to what's here in the repository), but we can't rely on it since it's not clear which version of R-INLA it provides, how it was built, on which OS, etc.

**Comment by finnlindgren** [comment]
The GNUmakefile in the build64.tgz file is essentially a newer version of the build-user linux makefile. I started working on re-integrating the feautres into the version in the repository, but haven't had time to finish that. In both cases, the file is mostly a wrapper for the individual makefiles in the inlaprog and gmrflib folders of the repository, plus extra compiler flags for finding the libraries and so on.
I also started looking at adding a cmake system to help the builds and clean up the configurations, but again, I haven't had time to work on that recently.

**Comment by finnlindgren** [comment]
I have a modified version of the build64.tgz makefiles that works more locally to build from a local clone of the gihub repository instead of downloading from github, to enable development testing on my own system. It would be nice to have more of this in the repository itself, to ensure the make system stays in sync with the code, and to support rebuilding older versions more easily, but I haven't had time to work on this in quite a while.

**Comment by hrue** [comment]
On Wed, 2021-04-07 at 10:44 -0700, Kenneth Hoste wrote:
> If build-user/linux/Makefile is not the latest version of the build
> procedure, is there another script/Makefile included in the repository
> that describes all the steps?

its all there. this is the one I use to compile on Fedora33, and they
are all the same up to some compiler options. 

For mac its the same and also windows (which I cross compile from Linux)






> We can check the contents of hrue.r-inla-download.org/build64.tgz for
> inspiration (if it includes useful stuff next to what's here in the
> repository), but we can't rely on it since it's not clear which
> version of R-INLA it provides, how it was built, on which OS, etc.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

-- 
Håvard Rue
Professor of Statistics
Statistics Program, CEMSE Division
King Abdullah University of Science and Technology
Thuwal 23955-6900, Saudi Arabia

***@***.***
Office: +966 (0)12 808 0640  
Mobile: +966 (0)54 470 0421
Research group: bayescomp.kaust.edu.sa  
R-INLA project: www.r-inla.org
Zoom: kaust.zoom.us/my/haavard.rue

**Comment by terjekv** [comment]
As a continuation of this topic, I need to be able to build R-INLA with PARDISO support from source, as the binaries fail on RHEL8 due to its binary build targeting libraries not on the system:

If I try to use the binary version of INLA_22.05.07 (upgraded from 21.05.02) under R-4.0.4, I get: 
````
> inla.pardiso.check()
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /tmp/terjekv/tmp-installs/R-INLA-21.05.02-
foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libcrypto.so.1.1: version `OPENSSL_1_1_1b' not found (required by /lib64/libk5crypto.so.3)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /tmp/terjekv/tmp-installs/R-INLA-21.05.02-
foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libcrypto.so.1.1: version `OPENSSL_1_1_1b' not found (required by /lib64/libk5crypto.so.3)
character(0)
attr(,"status")
[1] 1
Warning message:
In system(paste(shQuote(inla.call.no.remote()), "-m pardiso"), intern = TRUE) :
  running command ''/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl.run' -m pardiso' had status 1
````

If I try upgrade to the testing version INLA_22.12.12-2 I get: 
````
> inla.pardiso.check()
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/first/libRmath.so.1)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/first/libR.so)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libgsl.so.23)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/first/libpardiso.so)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libgfortran.so.5)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libicuuc.so.66)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libicui18n.so.66)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libmvec.so.1)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/first/libRmath.so.1)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/first/libR.so)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libgsl.so.23)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/first/libpardiso.so)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libgfortran.so.5)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libicuuc.so.66)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libicui18n.so.66)
/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/libmvec.so.1)
character(0)
attr(,"status")
[1] 1
Warning message:
In system(paste(shQuote(inla.call.no.remote()), "-m pardiso"), intern = TRUE) :
  running command ''/tmp/terjekv/tmp-installs/R-INLA-21.05.02-foss-2020b-R-4.0.4.eb/software/R-INLA/21.05.02-foss-2020b-R-4.0.4/INLA/bin/linux/64bit/inla.mkl.run' -m pardiso' had status 1
````

Would it be possible to provide instructions for adding PARDISO to the building of R-INLA from source?

**Comment by barracuda156** [comment]
> On Wed, 2021-04-07 at 10:44 -0700, Kenneth Hoste wrote: If build-user/linux/Makefile is not the latest version of the build procedure, is there another script/Makefile included in the repository that describes all the steps?
> its all there. this is the one I use to compile on Fedora33, and they are all the same up to some compiler options. For mac its the same and also windows (which I cross compile from Linux)
> […](#)

@hrue Should INLA be first built outside of R and then linked into R package? Asking, since `rinla` has no sources in it.
(Using any binaries is not acceptable option.)

**Comment by finnlindgren** [comment]
@barracuda156 Yes, one has to first build the binaries, place them in the needed folders (the build scripts I use do that, and I think the ones discussed by @hrue does that as well) and then build the R package.

**Comment by NewUser36** [comment]
Hello @finnlindgren and @hrue, I am trying to install INLA, and unfortunately since the package is not available on CRAN, the only way I can install INLA on my computer is by downloading the source code and compiling it myself.

I am new to compiling packages and I would like to know if there is a step-by-step guide I could follow on how to compile R-INLA in order to use the library in R. I have looked online but I haven't been able to make anything work (I'm not sure which makefile I should use, or something else entirely...)

I have downloaded the zip file for the code at this commit : https://github.com/hrue/r-inla/tree/dabdf621046e6fcb22d818d288c679bdb2c938ee. Which makefile should I use to compile? Is it the one in build-user/linux, or the one in the first available folder?

I am on Fedora 37 and R 4.2.3.

Thanks!

**Comment by hrue** [comment]
You can download as described in r-inla.org

After install, you can do

library(INLA)
inla.binary.install()

once only after each upgrade, to chose Fedora 37 builds, which are then
downloaded and installed.



On Tue, 2023-05-02 at 05:19 -0700, NewUser36 wrote:
> Hello @finnlindgren, I am trying to install INLA, and unfortunately
> since the package is not available on CRAN, the only way I can install
> INLA on my computer is by downloading the source code and compiling it
> myself.
> I am new to compiling packages and I would like to know if there is a
> step-by-step guide I could follow on how to compile R-INLA in order to
> use the library in R. I have looked online but I haven't been able to
> make anything work (I'm not sure which makefile I should use, or
> something else entirely...)
> I have downloaded the zip file for the code at this commit :
> https://github.com/hrue/r-inla/tree/dabdf621046e6fcb22d818d288c679bdb2c938ee
> . Which makefile should I use to compile? Is it the one in build-
> user/linux, or the one in the first available folder?
> I am on Fedora 37 and R 4.2.3.
> Thanks!
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>

-- 
Håvard Rue
Professor of Statistics
Chair of the Statistics Program
CEMSE Division
King Abdullah University of Science and Technology
Thuwal 23955-6900
Kingdom of Saudi Arabia

***@***.***
Office: +966 (0)12 808 0640  
Mobile: +966 (0)54 470 0421
Research group: bayescomp.kaust.edu.sa  
R-INLA project: www.r-inla.org
Zoom: kaust.zoom.us/my/haavard.rue

-- 

This message and its contents, including attachments are intended solely 
for the original recipient. If you are not the intended recipient or have 
received this message in error, please notify me immediately and delete 
this message from your computer system. Any unauthorized use or 
distribution is prohibited. Please consider the environment before printing 
this email.

**Comment by NewUser36** [comment]
Thanks for your answer, but downloading as explained in [https://www.r-inla.org/download-install](url) is not an option, since I do not have access to internet from rstudio on my work computer, which is why I wanted to compile and build everything myself.

**Comment by hrue** [comment]
If you do not have access to internet from rstudio, then you can
download at some other location, 

https://inla.r-inla-download.org/R/stable/src/contrib/INLA_23.04.24.tar.gz

the default binary build is Ubuntu-2004, which will also run on Fedora.

Then you just do

	R CMD INSTALL ./INLA_23.04.24.tar.gz

from the cmd line




On Tue, 2023-05-02 at 05:37 -0700, NewUser36 wrote:
> Thanks for your answer, but downloading as explained in
> https://www.r-inla.org/download-install is not an option, since I do
> not have access to internet from rstudio on my work computer, which is
> why I wanted to compile and build everything myself.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>

-- 
Håvard Rue
Professor of Statistics
Chair of the Statistics Program
CEMSE Division
King Abdullah University of Science and Technology
Thuwal 23955-6900
Kingdom of Saudi Arabia

***@***.***
Office: +966 (0)12 808 0640  
Mobile: +966 (0)54 470 0421
Research group: bayescomp.kaust.edu.sa  
R-INLA project: www.r-inla.org
Zoom: kaust.zoom.us/my/haavard.rue

-- 

This message and its contents, including attachments are intended solely 
for the original recipient. If you are not the intended recipient or have 
received this message in error, please notify me immediately and delete 
this message from your computer system. Any unauthorized use or 
distribution is prohibited. Please consider the environment before printing 
this email.
